### PR TITLE
Debian Packaging Update

### PR DESCRIPTION
**Pull request by hlieberman**
@fmarier, I've updated the Debian packaging over the last few days, based off of the git history as of dd4b8e87cbddff68273104825aabef0b34f9b284.

I broke it out into multiple binary packages, verified all the tests run clean, and tested the installation.  Everything seems to work.  Lintian is complaining mostly about expected things - UNRELEASED, as well as a couple of manpage errors - but nothing that's too unexpected.  I spent a couple of days trying to use the standard toolset to get this to build properly - pybuild, etc., but ended up having to go with a more manual process.  I squashed a lot of my attempts down in this branch, but an alternate strategy is still extant on my branch at hlieberman/debian-failed-attempt.

Please let me know if this looks good to you, and upstream can merge after your thumbs up.

### CODE REVIEW

**Comment by hlieberman** [comment]
Oh, and the debian branch is probably going to have to be force merged this time, because it's not based off of the previous branch - though I used the files as a starting point.  In the future, gitpkg should mean that we can just do fast-forward merges (most of the time).

File: debian/README.source [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+methodology.  Patches normally kept in debian/patches are instead
+broken out as branches based off of the upstream (master) branch, and
+are automatically regenerated at the time of the .dsc build by running
+`gitpkg debian` (or `gitpkg debian upstream` to regenerate the .orig
+tarball).
+
+As upstream is not yet giving us releases, we are using the version
+0.<date of retrieval from git>-<debian revision>.  We will coordinate
+with upstream before upload to sid or experimental to ensure that
+their first version will be > 0.2, or change our version format to
```
Comment (kuba) [inline debian/README.source]: Is there a specific requirement for `>0.2` or is that a typo? Could the first release be versioned at `0.1`?

File: debian/letsencrypt-doc.doc-base [hunk 2]
```diff
@@ -0,0 +1,12 @@
+Document: letsencrypt-manual
```
Comment (kuba) [inline debian/letsencrypt-doc.doc-base]: `letsencrypt-manual` might be easily confused with the manual authentication plugin (embedded in the core, rather than being distributed in a separate PyPI package). Could we have another name here?

File: debian/watch [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# The project has no releases, yet, so I am incrementing the patch
+# version manually and using things directly from git.
```
Comment (kuba) [inline debian/watch]: I guess you could count the number of commits as a stable increasing version component and for the `watch` purposes.

File: debian/rules [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	cd $(CURDIR)/letsencrypt-apache; \
+	python2.7 setup.py build; \
+	python2.7 setup.py install --single-version-externally-managed --root=$(CURDIR)/debian/letsencrypt-apache --install-layout=deb
+	# letsencrypt-nginx
+	cd $(CURDIR)/letsencrypt-nginx; \
+	python2.7 setup.py build; \
+	python2.7 setup.py install --single-version-externally-managed --root=$(CURDIR)/debian/letsencrypt-nginx --install-layout=deb
+	# Documentation
+	sphinx-build -N -b html docs $(CURDIR)/debian/doc-build/html
+	sphinx-build -N -b man docs $(CURDIR)/debian/doc-build/man
```
Comment (kuba) [inline debian/rules]: Note, this includes documentation for core, as well as ACME module, Apache and Nginx plugins.

File: debian/control [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	       python-zope.interface
+Standards-Version: 3.9.6
+Homepage: https://www.letsencrypt.org/
+X-Python-Version: >= 2.7
+
+Package: python-acme
+Section: python
+Architecture: all
+Depends: openssl,
+	 python-configargparse,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+               python-zope.component,
+	       python-zope.interface
+Standards-Version: 3.9.6
+Homepage: https://www.letsencrypt.org/
+X-Python-Version: >= 2.7
+
+Package: python-acme
+Section: python
+Architecture: all
+Depends: openssl,
```
Comment (kuba) [inline debian/control]: I don't think openssl is required.

However, `ca-certificates` is required for communication with Let's Encrypt CA (might be put as dep in the core package).

File: debian/control [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ designed as part of the Let's Encrypt project, to make it possible to
+ setup an HTTPS server and have it automatically obtain a
+ browser-trusted certificate, without any human intervention.  This
+ library implements the protocol used for proving the control of a
+ domain.  It currently supports only Python 2.
+
+Package: letsencrypt
+Architecture: all
+Depends: dialog,
+	 openssl,
```
Comment (kuba) [inline debian/control]: openssl is not required. It can be put into `Suggests`, because manual authenticator uses it in the output (though this output might not be run on the same machine, but on the target webserver instead)

File: debian/control [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	 python-acme,
+         python-configargparse,
+         python-configobj,
```
Comment (kuba) [inline debian/control]: you are missing `python-cryptography>=0.7` here

File: debian/control [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+         python-jsonschema,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+         python-mock,
+         python-ndg-httpsclient,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+         python-openssl (>= 0.15),
+         python-parsedatetime,
+         python-psutil,
+         python-pyasn1,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+         python-pyparsing,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+         python-requests,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+         python-rfc3339,
+         python-tz,
+         python-werkzeug,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+         python-jsonschema,
+         python-mock,
+         python-ndg-httpsclient,
+         python-openssl (>= 0.15),
+         python-parsedatetime,
+         python-psutil,
+         python-pyasn1,
+         python-pyparsing,
+         python-requests,
+         python-rfc3339,
```
Comment (kuba) [inline debian/control]: you are missing `python-dialog` dependency here

File: debian/control [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+  - Help you revoke the certificate if that ever becomes necessary.
+ .
+ This package contains the standalone client.
+
+Package: letsencrypt-apache
+Architecture: all
+Depends:
+	apache2,
+	letsencrypt,
+	python-augeas,
```
Comment (kuba) [inline debian/control]: missing `python-zope.component`, `python-zope.interface`, `python-mock` and `python-acme`

File: debian/control [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+         python-requests,
+         python-rfc3339,
+         python-tz,
+         python-werkzeug,
+         python-zope.component,
+         python-zope.interface,
+         ${misc:Depends},
+         ${python:Depends}
+Suggests: letsencrypt-doc
+Description: Let's Encrypt standalone client
```
Comment (kuba) [inline debian/control]: "standalone" can be easily confused with Standalone Authenticator plugin. How about "Let's Encrypt client"?

File: debian/control [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+ the web server.
+ .
+ This agent is used to:
+ .
+  - Automatically prove to the Let's Encrypt CA that you control the website
+  - Obtain a browser-trusted certificate and set it up on your web server
+  - Keep track of when your certificate is going to expire, and renew it
+  - Help you revoke the certificate if that ever becomes necessary.
+ .
+ This package contains the standalone client.
```
Comment (kuba) [inline debian/control]: again, phrasing is confusing. We technically include `standalone` and `manual` plugins in the core.

File: debian/control [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+  - Automatically prove to the Let's Encrypt CA that you control the website
+  - Obtain a browser-trusted certificate and set it up on your web server
+  - Keep track of when your certificate is going to expire, and renew it
+  - Help you revoke the certificate if that ever becomes necessary.
+ .
+ This package contains the documentation.
+
+Package: letsencrypt-nginx
+Architecture: all
+Depends: letsencrypt, nginx, ${misc:Depends}, ${python:Depends}
```
Comment (kuba) [inline debian/control]: missing `python-acme`, `python-pyparsing`, `python-mock`, `python-zope.interface`

File: debian/letsencrypt.dirs [hunk 20]
```diff
@@ -0,0 +1 @@
+var/lib/letsencrypt
```
Comment (kuba) [inline debian/letsencrypt.dirs]: what about `/etc/letsencrypt` and `/var/logs/letsencrypt`?

**Comment by kuba** [comment]
Great work, @hlieberman! :) How could I test it (build the packages)? Can we put some instructions in the docs?

File: acme/acme/client.py [hunk 21]
```diff
@@ -0,0 +1,582 @@
...
+from acme import errors
+from acme import jose
+from acme import jws
+from acme import messages
+
+
+logger = logging.getLogger(__name__)
+
+# Python does not validate certificates by default before version 2.7.9
+if sys.hexversion < 0x20709F0 :
```
Comment (kuba) [inline acme/acme/client.py]: Nice find! Could you please send this update in a separate PR to `master`? Can we use `sys.version_info < (2, 7, 9)` instead?

File: acme/setup.py [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+from setuptools import setup
+from setuptools import find_packages
+
+
+install_requires = [
+    'argparse',
+    # load_pem_private/public_key (>=0.6)
+    # rsa_recover_prime_factors (>=0.8)
+    'cryptography>=0.8',
+    'mock',
```
Comment (kuba) [inline acme/setup.py]: I have a patch ready, where conditional version tag is applied for Python 2.6. This way `debian` branch doesn't have to diverge from `master`. Same for all other `setup.py` scripts.

File: debian/README.source (hunk 1 above)
Comment (hlieberman) [inline debian/README.source]: It's because right now it's 0.<date>, and it's 2015.  If you release it as 0.1 and we already put it in Debian with 0.2015xxxx, then we will need to use epochs to make sure the versioning is correct.  README.source is notes for the other Debian maintainers, not for users.

File: debian/letsencrypt-doc.doc-base (hunk 2 above)
Comment (hlieberman) [inline debian/letsencrypt-doc.doc-base]: This isn't exposed to the user.  The package name is "letsencrypt-doc".

File: debian/watch (hunk 3 above)
Comment (hlieberman) [inline debian/watch]: There's no real easy way to do that inside a watch file, unfortunately.  This isn't a big deal - the file being present just shuts up one of our lint programs.

File: debian/rules (hunk 4 above)
Comment (hlieberman) [inline debian/rules]: Is there a way to split it out easily?  If not, I don't see that as a big problem.  It's definitely not a problem for -apache and -nginx.  python-acme is a little bit more questionable, but it's still not a big problem.

File: debian/control (hunk 5 above)
Comment (hlieberman) [inline debian/control]: python-setuptools?  It's required by dh-python.

File: debian/control (hunk 6 above)
Comment (hlieberman) [inline debian/control]: Doesn't matter, really.  ca-certificates deps on openssl.  I'll add the dep on ca-certificates.

File: debian/control (hunk 5 above)
Comment (kuba) [inline debian/control]: no, `python-configargparse` - all my comments are below the affected line

File: debian/control (hunk 6 above)
Comment (kuba) [inline debian/control]: The point is that someone has to maintain and sync `setup.py` with those lists - putting unnecessary stuff, such as `openssl`, makes this process harder.

If there is no dependency, then don't add it. If there is a transitive dependency (such as `ca-certificates`), but still used explicitly by our package - add it, so that if upstream changes their deps list we are not affected.

File: debian/control (hunk 5 above)
Comment (kuba) [inline debian/control]: For some reason this diff appears as outdated, but has not been addressed in the code.

**Comment by fmarier** [comment]
First of all, thanks @hlieberman for getting this into shape, I can see you've solved quite a few issues! In fact, too many for Github to be able to display them properly :)

A few general comments:

- You should add yourself to `debian/copyright` :)
- `debian/rules` is kind of messy now, but we can fix this later using pybuild, no need to block on this.
- I'm not a fan of the way gitpkg keeps patches to upstream code. I would vote for using git-buildpackage so that any deviation from upstream is easy to spot and manage through the `debian/patches/` directory. It's also a more familiar tool for Debian maintainers.

Also, a question for **upstream**, given that we're getting close to GA, can we start tagging releases for real? No need to actually push them to pyPI or produce tarballs, but if we could start practicing the whole packaging lifecycle now from a git tag, it would make things easier and ensure everything is working before launch.

We discussed the versioning scheme in an issue I can no longer find, but I would suggest: 

- `major.minor.patch` (e.g. 0.1.1)
- always *increasing* (no "0.1.1beta1" or stuff like that)

The second point is important to support package upgrades in Debian (and I presume other distros too).

File: acme/setup.py (hunk 22 above)
Comment (kuba) [inline acme/setup.py]: Fixed in #708.

**Comment by hlieberman** [comment]
Hi @fmarier!

Glad to help.  I updated d/copyright to add myself.

I tried to get it working with pybuild, but because of the specifics with how the tests interact with the nested setup.py's, I was never able to quite get it working right.  I spoke with Piotr about it when I was at DebConf, as well as paultag, and they recommended that I try this approach instead.  If you can get it working with pybuild, that'd certainly be a lot cleaner than the stuff I had to do to get it to build!

I find it easier to keep track of the diffs personally with gitpkg, since everything's in git already and you can just compare branches. If you prefer git-buildpackage, though, I am happy to convert it over.

Let me know!

**Comment by fmarier** [comment]
> I tried to get it working with pybuild, but because of the specifics with how the tests interact with the nested setup.py's, I was never able to quite get it working right. I spoke with Piotr about it when I was at DebConf, as well as paultag, and they recommended that I try this approach instead. If you can get it working with pybuild, that'd certainly be a lot cleaner than the stuff I had to do to get it to build!

Well, if Piotr and Paul can't figure out how to do it, it's probably not doable, at least for now. Anyways, like I said, I don't think we should block on that.

> I find it easier to keep track of the diffs personally with gitpkg, since everything's in git already and you can just compare branches. If you prefer git-buildpackage, though, I am happy to convert it over.

I would prefer we stick to git-buildpackage. I don't expect we'll be carrying patches for very long given we have a great relationship with upstream ;)

**Comment by hlieberman** [comment]
I force updated the branch to rebase out the git-debcherry merges and added the patches to d/patches.

@fmarier, how does it look to you now?

@pde, is it possible for Francois and I to get commit access to only the debian branch?

**Comment by kuba** [comment]
Do you intend to address the review comments from above? :)

File: debian/control [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+ This package contains the documentation.
+
+Package: letsencrypt-nginx
+Architecture: all
+Depends:
+	letsencrypt,
+	nginx,
+	python-acme,
+	python-mock,
+	python-openssl,
```
Comment (kuba) [inline debian/control]: this is not required

File: debian/control [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+ domain.  It currently supports only Python 2.
+
+Package: letsencrypt
+Architecture: all
+Depends: dialog,
+	 openssl,
+	 python-acme,
+         python-configargparse,
+         python-configobj,
+	 python-cryptography (>= 0.8)
```
Comment (kuba) [inline debian/control]: this should be >= 0.7 as in comment above (https://github.com/letsencrypt/letsencrypt/pull/701/files#r37695655)

File: debian/control (hunk 7 above)
Comment (kuba) [inline debian/control]: still not fixed

File: debian/control [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+	 python-dialog,
+	 python-mock,
+         python-openssl (>= 0.15),
```
Comment (kuba) [inline debian/control]: >0.15 is not necessary

File: debian/control [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+         python-parsedatetime,
+         python-psutil,
```
Comment (kuba) [inline debian/control]: this, on the other hand, has to be `>=2.1.0`

File: debian/control [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+         python-requests,
```
Comment (kuba) [inline debian/control]: this is still not required

**Comment by kuba** [comment]
Please make sure that dependencies match up with `setup.py` scripts...

File: debian/control (hunk 27 above)
Comment (kuba) [inline debian/control]: correction, it's missing in `setup.py`; thanks for catching, PR on its way...

File: debian/control (hunk 23 above)
Comment (kuba) [inline debian/control]: correction: it is required, but not in setup.py

File: debian/control (hunk 8 above)
Comment (kuba) [inline debian/control]: continued in https://github.com/letsencrypt/letsencrypt/pull/701/files#r38389006

File: debian/control [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+         python-rfc3339,
+         python-tz,
+         python-zope.component,
+         python-zope.interface,
+         ${misc:Depends},
+         ${python:Depends}
+Suggests: letsencrypt-doc, openssl
```
Comment (kuba) [inline debian/control]: ah, so you put `openssl` into `Suggests`, but forgot to remove from `Depends` ;)

**Comment by kuba** [comment]
I think it's good to be merged once version pins and `Suggests`/`Depends` comment is resolved.

Please remember to submit `0001-Make-version-check-for-Python-urllib3-injection-more.patch` as a separate PR against `master`.

Also, I would really love to test it out, but you never answered my question about instructions... :P

**Comment by fmarier** [comment]
> @fmarier, how does it look to you now?

This PR is really hard to review. For one thing, it's too big for GitHub. In the future, I think we need to submit the merge from master separately from the rest of the changes. That would make the diff view useful (in the second PR).

I've taken a look at the files in the debian/ directory and it seems like an improvement, so I'm OK with merging it and then fixing it up the missing things later (e.g. I think you dropped some remote image fixups for the docs in `debian/rules`).

**Comment by hlieberman** [comment]
Yeah, the PR is a bit messy.  What I'd suggest is doing `git log origin/debian..hlieberman/debian -p debian/` when you have my remote set as hlieberman.  That will give you the log, patch by patch, what changes I made on top of your last revision, with diffs, that only touch the debian directory.  It's still messy, since there's a lot of my work product in there in-line as I tried to get it to work, but it's easier.  The other option is just reviewing all of debian/ against the current merged version, since I didn't make any changes outside of it (that weren't later reverted when we switched away from git-debcherry).

Next time, it will be easier to parse, since there won't be all the work-product there.  Just the merge commit, and then a patch for the actual debian/ stuff.

I've pushed commits which fix-up a lot of cosmetic stuff - versions, whitespace, fixing other stuff like that.

**Comment by kuba** [comment]
LGTM, thanks

**Comment by hlieberman** [comment]
Remerged latest master in.

**Comment by jdkasten** [comment]
Sorry I missed the sign off on this by @fmarier.  I will merge it once conflicts are resolved.  If you don't have time @hlieberman , I will try to set aside some time tomorrrow.

Thanks @hlieberman, sorry again for the delay on this.

**Comment by hlieberman** [comment]
Merged with -s ours, since all the changes are either discarded or already in master.  Should be all set.

@kuba, if you want to test this, you should setup an sbuild chroot to build it in an isolated environment.  Once you have that done, you can just run `sbuild` in the main directory.  https://wiki.debian.org/sbuild will guide you through that process, though do note that if you follow those instructions, you will end up building packages for Debian sid (since that's where development is actually done).  If you want to build it under jessie, put "jessie" instead of "sid" everywhere in the sbuild-createchroot command and then build it with `sbuild -d jessie`.

**Comment by jdkasten** [comment]
Thanks @hlieberman!
